build-tools/packages/build-cli/src/commands/generate/sourceEntrypoints.ts

```diff
@@ -0,0 +1,386 @@
+
+		const mapSourceToExportPath: Map<ApiTag, ExportData> = getOutputConfiguration(
+			mapApiTagToSourcePath,
+			mapApiTagToExportPath,
+			this.logger,
+		);
+
+		if (mapSourceToExportPath.size === 0) {
+			throw new Error(
+				`There are no package exports matching requested output entrypoints:\n\t${[
+					...mapApiTagToSourcePath.keys(),
+				].join("\n\t")}`,
+			);
+		}
+
+		// generate source entrypoints under `${outDir}`
+		return generateSourceEntrypoints(mainEntrypoint, mapSourceToExportPath, this.logger);
+	}
+}
+
+// Formats the given output directory path to ensure it starts with `./` and ends with `/`.
+function formatPath(outDir: string): string {
+	return `./${outDir.replace(/^\.\/|\/$/g, "")}/`;
+}
+
+/**
+ * Read and parse tsconfig under `outDir`
+ * @returns rootDir
+ */
+async function readOutDirTsConfig(outDir: string): Promise<string> {
+	const formatOutDir = formatPath(outDir);
+
+	// Read tsconfig under `${outDir}`
+	const tsConfigContent = await fs.readFile(`${formatOutDir}tsconfig.json`, {
+		encoding: "utf8",
+	});
+
+	if (tsConfigContent === undefined) {
+		throw new Error(`tsconfig.json not found in ${formatOutDir}`);
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+	const tsconfig = JSON5.parse(tsConfigContent);
+
+	let rootDir = "";
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+	if (tsconfig.compilerOptions?.rootDir !== undefined) {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
+		rootDir = tsconfig.compilerOptions?.rootDir;
+	}
+
+	if (rootDir.length === 0 || rootDir === undefined) {
+		throw new Error(`No rootDir defined in ${formatOutDir}tsconfig.json`);
+	}
+
+	return formatPath(rootDir);
+}
+
+/**
+ * Returns a map of source paths to `ApiTag` levels.
+ */
+async function mapSourcePathToApiTag(outDir: string): Promise<Map<string, ApiTag>> {
+	const rootDir = await readOutDirTsConfig(outDir);
+
+	const outFileSuffix = ".ts";
+
+	return new Map([
+		[`${rootDir}${ApiLevel.alpha}${outFileSuffix}`, ApiTag.alpha],
+		[`${rootDir}${ApiLevel.beta}${outFileSuffix}`, ApiTag.beta],
+		[`${rootDir}${ApiLevel.public}${outFileSuffix}`, ApiTag.public],
+		[`${rootDir}${ApiLevel.legacy}${outFileSuffix}`, ApiTag.legacy],
+	]);
+}
+
+/**
+ * Resolves a mapping of `ApiTag` levels to their modified export paths.
+ */
+function getOutputConfiguration(
+	mapApiTagToSourcePath: ReadonlyMap<string, ApiTag>,
+	mapApiTagToExportPath: Map<ApiTag, ExportData>,
+	logger: CommandLogger,
+): Map<ApiTag, ExportData> {
+	const result = new Map<ApiTag, ExportData>();
+	for (const [sourcePath, apiTag] of mapApiTagToSourcePath) {
+		// Check if apiTag exists in mapApiTagToOutputPath
+		const exportPath = mapApiTagToExportPath.get(apiTag);
+		if (exportPath) {
+			// Modify the exportPath
+			const modifiedExportPath = exportPath.relPath
+				.replace(/(lib|dist)/g, "src")
```

Comment:
Replacement should be `outDir` to `rootDir` from tsconfig.

Reply:
In this case `outDir` is `./src/entrypoints/` and `rootDir` will possibly be the same. 

`relPath` is the export path (`./lib/entrypoints/public.js`, `./lib/entrypoints/beta.d.ts`)